### PR TITLE
Access to the encode/decode function From onPreExecute (2/2)

### PR DESCRIPTION
**Pull request by Ronan-Lenor**
Ah sorry, the 2 new function that i have provide recently can't be used in the onPreExecute Event.

So i had to entirely review the functionnality.

Now it is accessible on both onPreExecute & onPostExecute Event. AND it's much more easy to use.
As you don't have to provide the Request parameter in the method :P

### CODE REVIEW

**Comment by Ronan-Lenor** [comment]
part 1/2 : https://github.com/helios-ag/ElFinderPHP/pull/33

File: Loader/ElFinderLoader.php [hunk 1]
```diff
@@ -94,35 +94,60 @@ public function setConfigurator($configurator)
...
-     * @return string
+     *
+     * @return mixed
      **/
-    public function encode(Request $request, $path)
+    public function encode($path)
     {
-        $target = ($request->isMethod('POST')) ? $request->get('target') : $request->query->get('target');
+        if ($this->bridge === null) {
+            if (empty($this->instance)) {
```
Comment (nicolasmure) [inline Loader/ElFinderLoader.php]: The instance is setted when called the [`load`](https://github.com/helios-ag/FMElfinderBundle/blob/master/Controller/ElFinderController.php#L191) method on `ElFinderLoader`, right?
So when calling this encode function on a pre execution event will always throw an exception, right?

**Comment by Ronan-Lenor** [comment]
As you said, the bridge is set on  `load`  method.
so in `onPreExecute` the bridge would be never be loaded as the method `load` is not yet called.

if you wanna encode a path, you will just follow that step.
1. get the loader service
2. use `$loader->setInstance('myHomeFolder')`  (only if you are in `onPreExecute`)
3. `$loader->encode('my/encoded/path')`

the error is throw only if you forget to set the instance.
the 2nd step is useless in `onPostExecute` as all is already set.

**Comment by nicolasmure** [comment]
Yep, it might seem annoying to have to get the loader service inside a preExecution event listener and to set its instance, don't you think?
It could be easier to do something like this : 
```php
// FM\ElfinderBundle\Controller\ElFinderController
public function loadAction(Request $request, $instance, $homeFolder)
{
    $loader = $this->get('fm_elfinder.loader');
    $loader->initBridge($instance); // builds up the Bridge object for the loader with the given instance

    $httpKernel = $this->get('http_kernel');
    $preExecutionEvent = new ElFinderPreExecutionEvent($request, $httpKernel, $instance, $homeFolder);
    $this->get('event_dispatcher')->dispatch(ElFinderEvents::PRE_EXECUTION, $preExecutionEvent);
    
    $result = $loader->load($request); // the instance is already set
    
    $postExecutionEvent = new ElFinderPostExecutionEvent($request, $httpKernel, $instance, $homeFolder, $result);
    $this->get('event_dispatcher')->dispatch(ElFinderEvents::POST_EXECUTION, $postExecutionEvent);
    
    // returning result (who may have been modified by a post execution event listener)
    return new JsonResponse($postExecutionEvent->getResult());
}
```

```php
// FM\ElfinderBundle\Loader\ElFinderLoader
public function initBridge($instance)
{
    $this->setInstance($instance);
    $config = $this->configure();
    $this->bridge = new ElFinderBridge($config);
}

public function load(Request $request)
{
    $connector = new ElFinderConnector($this->bridge);

    if ($config['corsSupport']) {
        return $connector->execute($request->query->all());
    } else {
        $connector->run($request->query->all());
    }
}
```
Now you don't have to care if the Bridge attribute on the Loader service is set or not in your encode and decode functions, and you don't have to set the instance neither in a preExecution event listener.

What do you think of it ?

**Comment by Ronan-Lenor** [comment]
Yup, it's cool like that too.
but just one question,  if you don't get the loader service in the listener, how do you access to encode/decode function?

**Comment by nicolasmure** [comment]
Oh sorry, maybe I misspoke but of course you'll have to access to the loader in execution event listeners, but you won't have to set the instance to this loader anymore inside the event listeners as it will already be configured.

**Comment by Ronan-Lenor** [comment]
well, all modification have been commit.
Hope it would be Ok this time :P

**Comment by Ronan-Lenor** [comment]
how come that Scrutinizer didn't work?

**Comment by Ronan-Lenor** [comment]
ah! wait, i have notice an error in your code nicolasmure.
i will correct it

**Comment by nicolasmure** [comment]
well seen :sunglasses: !

**Comment by Ronan-Lenor** [comment]
No Cognac today :beer: :smile:  haha

**Comment by helios-ag** [comment]
I think you need to squash your commits before we merge.

**Comment by Ronan-Lenor** [comment]
to squash my commits?
sorry, i don't understand what you mean.

**Comment by helios-ag** [comment]
Useful information can be found here https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

**Comment by Ronan-Lenor** [comment]
Any idea of how can i do a rebase on the github website?

**Comment by nicolasmure** [comment]
The main of squashing your commits is to regroup them together.
As you have 22 commits here, just run
```bash
git rebase -i HEAD~22
```
then replace all the `pick` by `squash` except for the 1st one. It will regroup all your commits into the 1st one.
Save the git dialog ; then git prompts you about the comments you want to keep : write an explicit comment that illustrates your PR.
Save again and then
```bash
git push origin master --force
```
to push your changes. --forces is mandatory here as you are breaking the git tree with a rebase action.

**Note:** apparently, you're working directly in the master branch of your fork, I advise you to always work in dedicated branches instead of making changes directly in the master branch.

**Comment by Ronan-Lenor** [comment]
ah, i can't close that PR here anymore?

well, i have create a new one here: https://github.com/helios-ag/FMElfinderBundle/pull/176

**Comment by helios-ag** [comment]
:stuck_out_tongue_closed_eyes:

**Pull request closed without merging**